### ISSUE TITLE
Multiple trigger on `onDownloadProgress` callback

### ISSUE THREAD

**Issue by rtritto** [theogravity/electron-dl-manager#7]
`onDownloadProgress` callback is triggered 3 times (3 `onDownloadProgress.progress` logs) when the `percentCompleted` became _100_.

It should be trigged only once.

```ts
ipcMain.on('download', async (event: Electron.IpcMainInvokeEvent, downloadUrl: string) => {
  await manager.download({
    window: mainWindow,
    url: downloadUrl,
    saveDialogOptions: {},
    callbacks: {
      onDownloadProgress: ({ ipercentCompleted }) => {
        console.log('onDownloadProgress.progress: ', percentCompleted);
        mainWindow.webContents.send('download-progress', percentCompleted)
      },
      onDownloadCompleted: ({ item }) => {
        mainWindow.webContents.send('download-completed', item.getFilename())
      }
    }
  })
})
```

### Actual
```log
...
onProgress.progress:  95.06
onProgress.progress:  98.07
onProgress.progress:  100
onProgress.progress:  100
onProgress.progress:  100
```

### Expected

```log
...
onProgress.progress:  95.06
onProgress.progress:  98.07
onProgress.progress:  100
```

**Comment by theogravity** [comment]
This can happen in some situations, for example using data URLs, where the downloaded bytes > total bytes as reported directly from electron. 

https://github.com/theogravity/electron-dl-manager/blob/main/src/DownloadInitiator.ts#L234

In this situation, we keep it at 100 vs reporting a > 100 %

https://github.com/theogravity/electron-dl-manager/blob/main/src/utils.ts#L107

We listen to the updated event from electron here:

https://github.com/theogravity/electron-dl-manager/blob/main/src/DownloadInitiator.ts#L168

It is probably bug on the electron side probably relating to how the current / total bytes are calculated (I'm guessing in a data-uri situation, it might use the base64 encoding for total bytes, but then when expanded to binary, it's a different value for current), and because of this mismatch, it will continue to fire progress updates.

**Comment by theogravity** [comment]
It looks like also at the chromium level, this can bug out too in certain situations:

https://source.chromium.org/chromium/chromium/src/+/main:components/download/internal/common/download_item_impl.cc;l=1099;drc=071a85757d9d95316f83f50e1fe087a6d87908f9

**Comment by theogravity** [comment]
I've submitted a PR to the electron project to add more native methods on the `DownloadItem` that should hopefully give better fidelity (I can reference the internal % progress instead of calculate it myself). It may or may not fix your issue.

Once that's in, I'll update the references in this lib to use it.

https://github.com/electron/electron/pull/42805

**Comment by theogravity** [comment]
Good news, the PR has been merged in. I'll have to wait for the backports to also merge for older electron versions (30, 31, 32). Once they're in, I'll have a specific check to see if you're using electron >= 30 and apply the native calculation methods instead.

**Comment by theogravity** [comment]
Published under `3.1.0`. You'll need Electron version `30.3.0` or greater to get the native reporting stats.

Please re-open if it's still an issue after using the supported Electron version.
